### PR TITLE
User Story 1556887: [MIEngine Natvis] Support size specifiers for pointers as arrays

### PR DESCRIPTION
**Pull request by gc46**
New test for array pointers is still failing.

### CODE REVIEW

File: test/CppTests/debuggees/natvis/src/visualizer_files/Simple.natvis [hunk 1]
```diff
@@ -31,6 +31,16 @@
   </Expand>
   </Type>
 
+  <Type Name="SimpleArrayPointer">
```
Comment (WardenGnaw) [inline test/CppTests/debuggees/natvis/src/visualizer_files/Simple.natvis:34]: There should be no need for a natvis type for testing the `<variableName>,<size>` format. 

I would recommend testing on a regular int pointer pointing to a block of memory then evaluating the format specifier on it.

File: test/CppTests/Tests/NatvisTests.cs [hunk 2]
```diff
@@ -330,6 +330,47 @@ public void TestThisConditional(ITestSettings settings)
...
+                SourceBreakpoints writerBreakpoints = debuggee.Breakpoints(NatvisSourceName, ReturnSourceLine);
+                runner.SetBreakpoints(writerBreakpoints);
+
+                runner.Expects.StoppedEvent(StoppedReason.Breakpoint).AfterConfigurationDone();
+
+                using (IThreadInspector threadInspector = runner.GetThreadInspector())
+                {
+                    IFrameInspector currentFrame = threadInspector.Stack.First();
+
+                    this.Comment("Verifying 'ArrayPointer' natvis");
```
Comment (WardenGnaw) [inline test/CppTests/Tests/NatvisTests.cs:364]:
```suggestion
                    this.Comment("Verifying comma format specifier");
```

File: test/CppTests/Tests/NatvisTests.cs [hunk 3]
```diff
@@ -330,6 +330,47 @@ public void TestThisConditional(ITestSettings settings)
...
+            this.TestPurpose("This test checks if ArrayPointer is visualized.");
+            this.WriteSettings(settings);
+
+            IDebuggee debuggee = Debuggee.Open(this, settings.CompilerSettings, NatvisName, DebuggeeMonikers.Natvis.Default);
+
+            this.Comment("Run the debuggee, check argument count");
+            using (IDebuggerRunner runner = CreateDebugAdapterRunner(settings))
+            {
+                this.Comment("Configure launch");
+                string visFile = Path.Join(debuggee.SourceRoot, "visualizer_files", "Simple.natvis");
```
[the remark this answers is not in the document]
Comment (gc46) [reply test/CppTests/Tests/NatvisTests.cs]: I think this is resolved?

**Review by WardenGnaw — APPROVED** [review]

**Comment by WardenGnaw** [comment]
osx test failure is flaky.

File: src/MIDebugEngine/Engine.Impl/Variables.cs [hunk 4]
```diff
@@ -431,7 +431,13 @@ private string StripFormatSpecifier(string exp, out string formatSpecifier)
             // TODO: could return '(T(*)[n])(exp)' but requires T
             var m = Regex.Match(trimmed, @"^\[?(\d+)\]?$");
             if (m.Success)
+            {
+                if (_engine.DebuggedProcess.MICommandFactory.Mode == MIMode.Gdb)
+                    return "*" + exp.Substring(0, lastComma) + "@" + trimmed;  // this does not work for lldb
```
Comment (KindDragon) [inline src/MIDebugEngine/Engine.Impl/Variables.cs:436]: @gc46 should work something like that `(lldb) parray 10 ptr`
https://lldb.llvm.org/use/map.html

Comment (WardenGnaw) [reply src/MIDebugEngine/Engine.Impl/Variables.cs:436]: We cannot use the print method of GDB or LLDB, MIEngine uses MI Commands to communicate well formed output.

However, LLDB support has been included in https://github.com/microsoft/MIEngine/pull/1345

**Pull request merged**